Make config options and GUI settings match

CyrilleB79 opened the pull request:
### Link to issue number:
Closes #14170
### Summary of the issue:
When setting line indentation reporting on tone in a profile (e.g. Notepad) and on speech for default profile, you end up with line indentation set on Speech and tones instead of tones only when turning back to Notepad (see #14170 for detailed steps).

The issue is due to the following:
Only one combo-box allows to set line indentation reporting in the UI. But the value of this combo-box is controlled by two items in the configuration. These two items are then managed separately when dealing with configuration profile switching.

The following combo-boxes in the GUI are impacted by this issue:
* "Line indentation reporting" in doc formatting settings (see initial description of #14170)
* "Cell borders" in doc formatting settings (see https://github.com/nvaccess/nvda/issues/14170#issuecomment-1277702879)
* "Show messages" combined with "Message timeout" in braille settings (see https://github.com/nvaccess/nvda/issues/14170#issuecomment-1276335227)
* "Tether Braille" in braille settings (see https://github.com/nvaccess/nvda/issues/14170#issuecomment-1277685893)

### Description of user facing changes

The values of the following settings should not change unexpectedly anymore when switching profile:
* Line indentation in Document formatting settings.
* Cell borders in doc formatting settings
* Show messages in braille settings
* Tether Braille in braille settings

### Description of development approach
* One control in the GUI now matches one and only one item in the config
* Upgrade the config adequately
* Used the same approach as #14132 (and the fix in #14183)

#### A note on profile upgrade
In some cases, the config does not allow to know the selected option. Let's take as an example the following config: reportIndent = True and reportIndentWithTones not present in the config.
1. If the config is for default profile, we are sure that the selected option in GUI for line indent is "with speech".
2. If this config is for another profile (profile "foo") stacked on default profile, the GUI line indent option also depends on the option selected in default profile:
   * if line indent in default profile is disabled, the line indent GUI option in foo profile will be "with speech"
   * if line indent in default profile is set on "With tones", the line indent GUI option in foo profile will be "with both speech and tones"

In such situations where one cannot exactly know the option selected in the GUI, the implementation uses the following rules by order of priority:
* assume that the user has modified the config only via the GUI (not via commands in Python console)
* assume first that the upgraded profile is default profile; only if the upgraded config is not possible in default config it is assumed that it can be for another profile.

### Testing strategy:
Manual tests:
For each combobox:
- Tested that each option works
- Tested (with a test profile) that config is correctly upgraded for each possible option.
- when applicable, ("line indent", "cell borders" and "tether to") tested the corresponding cycle/toggle script
- Tested the STR of the corresponding issue in #14170.

Unit tests added for each GUI option selected from default profile as well as for other corner cases.

### Known issues with pull request:
#### Issue 1: potential issues with the upgrade of non-default profiles
Some issues may remain when upgrading NVDA with profiles. Below is an example on a fresh NVDA:
* Set line indent to "speech" in default profile
* Set line indent to "both speech and tones" in a separate test profile
Before upgrade, default profile's config will contain the following `[documentFormatting]` section:
```
[documentFormatting]
	reportLineIndentation = True
```
Before upgrade, test profile's config will contain the following `[documentFormatting]` section:
```
[documentFormatting]
	reportLineIndentationWithTones = True
```
Although indentation is reported by speech and tones in test profile, `reportLineIndentation = True` is not specified in test profile's config file since it is inherited from default profile.

When upgrading, the test profile will have line indent set on "tones" instead of speech and tones.
This is because profiles are upgraded one at a time, but their config is defined according to the config on which they are stacked.

Should I indicate somewhere the following information?

> If one of the following parameters have been modified in non-default profiles:
> * Line indentation in Document formatting settings.
> * Cell borders in doc formatting settings
> * Show messages in braille settings
> * Tether Braille in braille settings
> in some rare cases, the value of these parameters may be modified unexpectedly when transitioning from NVDA 2022.4 to 2023.1. Please check the configuration of your profiles.

#### Issue 2: changes for dev / API / deprecation
* should removal of fields in the config be considered API-breaking? I guess yes, but this has not been the case before (maybe by mistake?)
* in any case, should these changes in the config be specified in the change for developers?

### Change log entries:
Bug fixes
```
- Some options are not modified unexpectedly in a profile anymore when modifying another profile in some rare cases. (#14170)
  - In some rare cases, the following options may still be unexpectedly modified when installing this version of NVDA for the following options in profiles: Line indentation and Cell borders in Document formatting settings, Show messages and Tether Braille in braille settings. Please check these options in your profiles with this new version of NVDA.
```
Deprecation:
```
- ``braille.BrailleHandler.TETHER_*`` are deprecated. Please use ``configFlags.TetherTo.*.value`` instead.  (#14233)
```

API breaking changes:
```
- Some keys in the configuration have been removed or modified (#14233).
  - In ``[documentFormatting]`` section:
    - ``reportLineIndentation`` stores an int value (0 to 3) instead of a boolean and ``reportLineIndentationWithTones`` has been removed.
    - ``reportBorderStyle`` and ``reportBorderColor`` have been removed and are replaced by ``reportCellBorders``.
  - In ``[braille]`` section:
    - ``noMessageTimeout`` has been removed and ``messageTimeout`` cannot take the value 0 anymore; these use cases are replaced by values in ``showMessages``.
    - ``autoTether`` has been removed; ``tetherTo`` can now take the value "auto" instead.
```
### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

CyrilleB79 committed:
The following option with a combo-box in NVDA settings window is now controlled by only one parameter in the config:
Line indentation in Document settings.

CyrilleB79 committed:
Single config option matching show message option in braille settings.

CyrilleB79 committed:
Single config option matching Cell borders option in doc formatting settings.

CyrilleB79 committed:
Single config option matching Tether to option in Braille settings

seanbudd commented:
It looks like `include/sonic` was mistakenly changed in this PR

seanbudd commented on source/config/configSpec.py:
```diff
@@ -68,12 +68,12 @@
-	messageTimeout = integer(default=4,min=0,max=20)
-	tetherTo = string(default="focus")
-	autoTether = boolean(default=true)
+	messageTimeout = integer(default=4,min=1,max=20)
```
```suggestion
	messageTimeout = integer(default=4, min=1, max=20)
```

seanbudd replied:
Can you make this change as well

seanbudd commented on source/braille.py:
```diff
@@ -2028,7 +2034,7 @@ def message(self, text):
 		@postcondition: The message is displayed.
 		"""
-		if not self.enabled or config.conf["braille"]["messageTimeout"] == 0 or text is None:
+		if not self.enabled or config.conf["braille"]["showMessages"] == ShowMessages.DISABLED or text is None:
```
new lines for readability

```suggestion
		if (
			not self.enabled
			or config.conf["braille"]["showMessages"] == ShowMessages.DISABLED
			or text is None
		):
```

seanbudd commented on source/braille.py:
```diff
@@ -1818,7 +1824,7 @@ def setTether(self, tether, auto=False):
 		self.mainBuffer.clear()
 
 	def _get_shouldAutoTether(self):
```
```suggestion
	def _get_shouldAutoTether(self) -> bool:
```

seanbudd commented on source/config/configSpec.py:
```diff
@@ -68,12 +68,12 @@
-	tetherTo = string(default="focus")
-	autoTether = boolean(default=true)
+	messageTimeout = integer(default=4,min=1,max=20)
+	tetherTo = option("auto", "focus", "review", default="auto")
```
could this get turned into an Enum? is a FeatureFlag more appropriate for this (or any other) of these?

CyrilleB79 replied:
There are already `BrailleHandler.TETHER_*` values. I did not want to change this part of code just to convert it to an enum as it belongs to the interface and may be used by add-ons. Of course, 2023.1 is an API-breaking release; but IMO it's not worth modifying the API just for the beauty of the code.

Regarding featureFlags, if I understand the concept correctly, it's grossly an enum with a default value. It's adapted for  features in which we expect the default to be changed in future versions of NVDA (e.g. new features being tested). I do not imagine that "Tether to" option will be modified in the foreseeable future.

Let me know if you still want me to convert `BrailleHandler.TETHER_*` to an enum. Or if you just want me to introduce `BrailleHandler.TETHER_*` in the configSpec instead of hard-coded values (I do not know how).

CyrilleB79 replied:
Done in faf36b3c204de23673bb2c46964f11cf641fd172.

seanbudd commented on source/globalCommands.py:
```diff
@@ -594,28 +594,13 @@ def script_toggleReportLineNumber(self,gesture):
 		category=SCRCAT_DOCUMENTFORMATTING
 	)
 	def script_toggleReportLineIndentation(self,gesture):
```
```suggestion
	def script_toggleReportLineIndentation(self, gesture: inputCore.InputGesture):
```

CyrilleB79 replied:
I will apply your suggestions regarding typing in function signatures. Note however that these lines were not initially modified by this PR.

seanbudd replied:
considering the function was fully rewritten, I think it was worth doing. Thanks for accepting the changes.

seanbudd commented on source/globalCommands.py:
```diff
@@ -701,25 +686,13 @@ def script_toggleReportTableCellCoords(self,gesture):
 		category=SCRCAT_DOCUMENTFORMATTING,
 	)
 	def script_toggleReportCellBorders(self, gesture):
```
```suggestion
	def script_toggleReportCellBorders(self, gesture: inputCore.InputGesture):
```

seanbudd reviewed (commented):
Thanks for picking this up and fixing this, looks pretty good

CyrilleB79 committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

CyrilleB79 committed:
Remove unintended modification of sonic library.

lukaszgo1 reviewed (commented):
Given some potential issues with upgrading of profiles have you considered modifying the upgrade steps, so that they accept profiles present lower in the stack as an additional parameter? While this introduces additional complexity breaking user config is hardly ideal, and we may encounter similar cases in the future, so solution of some kind would probably be needed later on.

CyrilleB79 committed:
Update function description.

seanbudd commented on source/config/profileUpgradeSteps.py:
```diff
@@ -134,3 +139,160 @@ def upgradeConfigFrom_7_to_8(profile: Dict[str, str]) -> None:
 			profile['documentFormatting']['reportTableHeaders'] = ReportTableHeaders.OFF.value
+
+
+def upgradeConfigFrom_8_to_9(profile: Dict[str, str]) -> None:
```
It appears to be relatively easy to do this
https://stackoverflow.com/a/53845083/8030743

Note, I haven't tested this and it would be good to place this type in a more general location
```py
from __future__ import annotations 
configProfileT = Dict[str, Union[str, "configProfileT"]]
```

seanbudd replied:
@CyrilleB79 - can you make this change?

CyrilleB79 replied:
profile is of type `configobj.ConfigObj`, which is a child class of `dict`.
I do not understand why we used `Dict[str, str])` for type hint, and why we need to specify a dedicated recursive type for type hints instead of using `configobj.ConfigObj` instead.

Maybe I missed something about type hints?

seanbudd replied:
Oh, I'm not sure why it was introduced as that to begin with then. Could you use the real type `configobj.ConfigObj`? Thanks for picking this up

CyrilleB79 commented:
Thanks @lukaszgo1 for the review.

> Given some potential issues with upgrading of profiles have you considered modifying the upgrade steps, so that they accept profiles present lower in the stack as an additional parameter?

The main problem is that the profiles stack is not always the same. For example, if you have created an application profile for Notepad and a manual profile. Then you upgrade NVDA to 2023.1, there are two scenarios:
* Scenario 1: while in notepad, you first activate the manual profile; then the manual profile would be upgraded taking into account the Notepad profile and the default profile.
* Scenario 2: while NOT in notepad, you first activate the manual profile; then the manual profile would be upgraded taking into account only the default profile.
Thus, the upgrade would depend on the situation when you have activated first the manual profile.

We could make the assumption that only the default profile was active, what probably covers the majority of the cases. But frankly adding more complexity to it just to cover a few upgrades and even not be sure that they will all be correctly covered is not worth it IMO.

@seanbudd your opinion on this?




> While this introduces additional complexity breaking user config is hardly ideal, and we may encounter similar cases in the future, so solution of some kind would probably be needed later on.

CyrilleB79 commented on source/config/profileUpgradeSteps.py:
```diff
@@ -134,3 +139,160 @@ def upgradeConfigFrom_7_to_8(profile: Dict[str, str]) -> None:
+		reportLineIndent = False
+	try:
+		reportLineIndentWithTones: str = profile['documentFormatting']['reportLineIndentationWithTones']
+		del profile['documentFormatting']['reportLineIndentationWithTones']
```
I agree that keeping the old keys in the config does not cause any problem.
But I prefer deleting the old keys to clarify things for developers/testers when reading the config.

Regarding backward compatibility, even if I do not remove the old keys, it is not guaranteed by this PR as written today. For example `profile['documentFormatting']['reportLineIndentation']` can be upgraded to a value of 3 (speech and tones) with schema version = 9 whereas it can only be True or False with schema version = 8. I would have to avoid reusing old key names to avoid this issue.

Moreover, backward compatibility of configuration has never been a requirement and it has always been discouraged to downgrade NVDA. If there is an evolution on this topic, it should be acted by NVAccess. @seanbudd could you clarify NVAccess' position?

seanbudd replied:
>Moreover, backward compatibility of configuration has never been a requirement and it has always been discouraged to downgrade NVDA. If there is an evolution on this topic, it should be acted by NVAccess. @seanbudd could you clarify NVAccess' position?

This is still true. I think deleting the keys is probably the safer option for the long term.

feerrenrut replied:
Config is backed up before being upgraded. The safest way to "downgrade" would be to restore one of those config-files.
However, there are many UX edge cases to consider. I don't think the value for users is worth the extra development complexity.

CyrilleB79 replied:
OK. So I do not modify anything.

CyrilleB79 committed:
Use enum instead of constants for TETHER_*

seanbudd commented on source/braille.py:
```diff
@@ -1752,18 +1756,10 @@ def formatCellsForLog(cells: List[int]) -> str:
-		# Translators: The label for a braille setting indicating that braille should be tethered to the review cursor.
-		(TETHER_REVIEW,_("to review"))
-	]
+	TETHER_AUTO = TetherTo.AUTO.value
```
```suggestion
	# TETHER_AUTO, TETHER_FOCUS, TETHER_REVIEW and tetherValues
	# are deprecated, but remain to retain API backwards compatibility
	TETHER_AUTO = TetherTo.AUTO.value
```

seanbudd commented:
@feerrenrut - what do you think about the known issues?

feerrenrut commented:
About the known issues:
### Issue 1: potential issues with the upgrade of non-default profiles
- Before upgrading, with the "test profile" enabled, indentation is reported with speech and beeps, after upgrading indentation is reported only with tones. Do I understand this correctly?
- If so, some more details on why this happens would be useful.
- There aren't unit tests for the upgrader, introducing tests to cover this situation would make it easier to resolve.
- Ideally the config upgrader could be extended to make this situation easy to handle.

### Issue 2: changes for dev / API / deprecation
> should removal of fields in the config be considered API-breaking? 

Yes. This should be considered an API-breaking change.

> in any case, should these changes in the config be specified in the change for developers?

Yes.

CyrilleB79 committed:
Fix range issue in config spec.

CyrilleB79 committed:
Modify formatting as per review comment.

CyrilleB79 commented:
@feerrenrut , thanks for your comments. Here are my answers:

> ### Issue 1: potential issues with the upgrade of non-default profiles
> * Before upgrading, with the "test profile" enabled, indentation is reported with speech and beeps, after upgrading indentation is reported only with tones. Do I understand this correctly?

Correct

> * If so, some more details on why this happens would be useful.

I have updated the description of this issue. Let me know if it is more clear.

> * There aren't unit tests for the upgrader, introducing tests to cover this situation would make it easier to resolve.

Do you mean unit tests for the following functions and with the various initial possible profiles?
* _upgradeConfigFrom_8_to_9_lineIndent
* _upgradeConfigFrom_8_to_9_cellBorders
* _upgradeConfigFrom_8_to_9_showMessages
* _upgradeConfigFrom_8_to_9_tetherTo

Please confirm so that I can start this work without risking to provide something unuseful or unexpected.

> * Ideally the config upgrader could be extended to make this situation easy to handle.

As explained in https://github.com/nvaccess/nvda/pull/14233#issuecomment-1282044096, for profiles with unspecified default values, the option in the GUI can take various values according to the current profile stack. There is no mean to be sure that the user will use the same profile stack when activating first the profile in the new version of NVDA, i.e. when the profile will be upgraded.


> ### Issue 2: changes for dev / API / deprecation
> > should removal of fields in the config be considered API-breaking?
> 
> Yes. This should be considered an API-breaking change.
> 
> > in any case, should these changes in the config be specified in the change for developers?
> 
> Yes.

I have updated the change log paragraph in the initial description. Let me know if it is OK.

CyrilleB79 committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

feerrenrut commented:
> Do you mean unit tests for the following functions and with the various initial possible profiles?

No. A layer below, testing the schema upgrade mechanism itself. Use a custom schema an config / profiles to demonstrate where it doesn't handle required use cases, then we can look at handling those. 

While I'll certainly welcome more unit tests to exercise the code in NVDA, the direction of this PR may need to change.

The comment `class ConfigManager` (`source/config/__init__.py`) provides a good description of the feature that makes this config upgrade hard to support:

``` txt
In addition to the base configuration, there can be multiple active configuration profiles.
Settings in more recently activated profiles take precedence,
with the base configuration being consulted last.
This allows a profile to override settings in profiles activated earlier and the base configuration.
A profile need only include a subset of the available settings.
Changed settings are written to the most recently activated profile.
```

If I understand this correctly, unless all values that we want to combine are defined in that profile (or it is the default profile), the profile could have different behavior depending on which other profile was activated first.

```
defaultProfile:
 - featureA = false
 - featureB = false
profileA:
 - featureA = true
profileB:
 - featureB = true
profileC:
 - featureA = false
 - featureB = true
```
- Activate `profileA`, result `featureA` is enabled, `featureB` is not.
- Activate `profileB` result: `featureB` is enabled, `featureA` is not.
- Activate `profileC` result: `featureB` is enabled, `featureA` is not.
- Activate `profileB` then `profileA`, result: `featureA` and `featureB` is enabled.
- Activate `profileA` then `profileC`, result: `featureB` is enabled, `featureA` is not.

Trying to combine these values into a single config option is not possible while also persisting the ability to stack profiles in any arbitrary order. Confirming this in a unit test would be helpful.

I'm inclined to question whether this behavior is obvious to users, and also whether this level of configurability is required.
It may even be that some of the referenced issues are due to this UX complexity. But these questions lead to much bigger changes to how config / profiles work. Is there another approach that resolves the linked issues?

CyrilleB79 committed:
Modify type hints.

CyrilleB79 commented:
@feerrenrut keep in mind that this PR aims to fix only the issue coming from the fact that one GUI option is controlled by two values in the config. This issue does not aim to fix the issues that could arise from stacking profiles in various orders.

I will try to make the required unit tests to clarify what happens during upgrade so that we can discuss it. So do you recommend to make various unit tests on `profileUpgrader._doConfigUpgrade`, testing the upgrade of various configs from SM8 to SM9? Is it correct?

CyrilleB79 committed:
Merge branch 'master' into comboboxAndConfigMatch

CyrilleB79 committed:
Fix system tests

CyrilleB79 committed:
Unit tests for _upgradeConfigFrom_8_to_9_showMessages

CyrilleB79 committed:
Fix bug found while developping unit tests.

CyrilleB79 commented:
@feerrenrut, I have developed unit tests for _upgradeConfigFrom_8_to_9_showMessages function in 5b3b4fc6cdeb3f43d04d0f602c6625eda6db2fd0 (and also found a bug thanks to it). Would you mind have a look to see if it helps understanding the various upgrade cases?

If yes, I will implement unit tests for the 3 other functions.

Many thanks.

CyrilleB79 committed:
Merge branch 'master' into comboboxAndConfigMatch

CyrilleB79 committed:
Some more unit tests (WIP)

CyrilleB79 committed:
All normal unit tests (for default profile) implemented.

CyrilleB79 committed:
Fix bug after unit tests implementation.

CyrilleB79 committed:
Add one more corner case unit test.

CyrilleB79 commented:
Finally, I have added all the unit tests for normal config cases, as well as two corner case of non-default config.

This PR is ready again to be reviewed.

seanbudd commented on tests/unit/test_config.py:
```diff
@@ -342,3 +349,329 @@ def test_CustomEnum_equalsOp(self):
 		self.assertTrue(defaultNever == CustomEnum.NEVER)
+
+
+def _loadProfile(configString):
```
```suggestion
def _loadProfile(configString: str) -> configobj.ConfigObj:
```

CyrilleB79 committed:
Add type hints.

CyrilleB79 commented:
@seanbudd I have added the type hint you suggested as well as other ones. Thanks.

CyrilleB79 commented:
Note: I have edited the initial description, especially to make appear the main item in the change log that was not appearing before due to bad markdown formatting.

feerrenrut commented on tests/unit/test_config.py:
```diff
@@ -351,21 +351,114 @@ def test_CustomEnum_equalsOp(self):
+"""
+		profile = _loadProfile(configString)
+		_upgradeConfigFrom_8_to_9_lineIndent(profile)
+		self.assertEqual(profile['documentFormatting']['reportLineIndentation'], 0)
```
An int enum for 0, 1, 2 would be helpful.

CyrilleB79 replied:
Do you mean using `config.configFlags.ReportLineIndentation.OFF.value` instead?

I was unsure if it was a good idea to use another part of NVDA's code to unit test this function.

If it's OK for you, I'll convert all the numerical values the same way.

feerrenrut replied:
Yes, this is ok.

CyrilleB79 replied:
Fixed in 559460c89c28be31100a623eac9c4a83f4741b2c. You can now review this PR again.

feerrenrut commented on tests/unit/test_config.py:
```diff
@@ -351,21 +351,114 @@ def test_CustomEnum_equalsOp(self):
+"""[documentFormatting]
+	reportBorderStyle = True
+	reportBorderColor = True
+"""
```
This string seems to be unnecessary.

CyrilleB79 replied:
> This string seems to be unnecessary.

Sorry, I cannot get from where you commented. The line numbers do not match the ones in the last commit.

Such strings are the config being tested in each function of the unit tests.

feerrenrut replied:
Hmm, strange. I'm now worried something went wrong when I reviewed this. The changes look quite different now! I'll re-review this tomorrow.

feerrenrut reviewed (changes requested):
Do you plan to add tests for ambiguous multi profiles situations?
If so, or it's already covered and I'm not seeing it, then it would be good to highlight the issue and explain how it can affect developers.

CyrilleB79 commented:
> Do you plan to add tests for ambiguous multi profiles situations? If so, or it's already covered and I'm not seeing it, then it would be good to highlight the issue and explain how it can affect developers.

No I do not plan to add more tests unless you ask to. I have only added two unit tests for the non-obvious cases where the upgrade is done on profiles that are assumed not to be the default one given the content of the profile's config:
* test_ManualProfile_CellBordersColorAndStyle
* test_ManualProfile_TetherToReview

It seems to me that further tests highlighting the issue with arbitrary stacking order of profiles were out of the scope of this PR (according to https://github.com/nvaccess/nvda/pull/14233#issuecomment-1286425938)

feerrenrut commented:
>It seems to me that further tests highlighting the issue with arbitrary stacking order of profiles were out of the scope of this PR (according to https://github.com/nvaccess/nvda/pull/14233#issuecomment-1286425938)

In that comment I was trying to say that the more important test is the arbitrary stacking order behavior. Because it is the more surprising aspect. That said, if you don't want to write these tests, or they are difficult to write, then we can create a new github issue to track it as a problem that affects developers.

CyrilleB79 committed:
Use enum instead of hard-coded values.

CyrilleB79 commented on source/config/configFlags.py:
```diff
@@ -6,7 +6,90 @@
+			ReportLineIndentation.OFF: _("Off"),
+			# Translators: A choice in a combo box in the document formatting dialog to report indentation
+			# with Speech.
+			ReportLineIndentation.SPEECH: pgettext('line indentation setting', "Speech"),
```
Yes. I have taken the strings as originally defined in `settingsDialogs.py`, so that nothing changes in the translation files.

Probably `pgettext` has been used for `speech` because the translation of this word can vary depending on the context.

What is the goal of your remark? We may use `pgettext` for the 4 options if you wish.

CyrilleB79 commented on source/config/configFlags.py:
```diff
@@ -6,7 +6,90 @@
 """Flags used for the configuration."""
```
What more are you expecting?
Actually, the config flag is generic so there isn't much more to say with respect to feature flag here.
On the opposite, the feature flag is more specific, i.e. it's more or less a config flag with a default option. So maybe an update in the feature flag file? (even if we are a bit off-topic of this PR)

Just let me know.

feerrenrut replied:
I want it to be clear to a developer that both modules exist, and that they have similar but different intentions. I think I'd suggest that feature flags should be used for new settings, since they provide a more intuitive way for a user to return to the default option.

CyrilleB79 replied:
I have written up something in 9edee7e5ce957edea5665ac6d7a2849f0c4848fc.

CyrilleB79 committed:
Address review comments.

CyrilleB79 committed:
Linting

feerrenrut committed:
update changes file for PR #14233

feerrenrut committed:
Merge branch 'origin/master'

CyrilleB79 commented on user_docs/en/changes.t2t:
```diff
@@ -36,6 +36,17 @@ What's New in NVDA
+  - Cell borders in doc formatting settings
+  - Show messages in braille settings
+  - Tether Braille in braille settings
+  - In some rare cases, the following settings used in profiles may be unexpectedly modified when installing this version of NVDA:
```
@feerrenrut 
Actually, it seems that I repeat the name of the options two times. Maybe I could just write as below and remove the second list of options?

```suggestion
  - In some rare cases, these settings used in profiles may be unexpectedly modified when installing this version of NVDA.
```

feerrenrut replied:
Ah, ok. I'll adjust this, how about this:
```
- Some settings will no longer be unexpectedly different when used with multiple profiles. (#14170)
  - The following settings have been addressed:
    - Line indentation in Document formatting settings.
    - Cell borders in doc formatting settings
    - Show messages in braille settings
    - Tether Braille in braille settings
  - In some rare cases, these settings used in profiles may be unexpectedly modified when installing this version of NVDA.
  - Please check these options in your profiles after upgrading NVDA to this version.
  -
```

CyrilleB79 replied:
That's OK. Thanks.

feerrenrut committed:
fixup changes entry
